Run CI against ubuntu 14.04

Empact opened the pull request:
In light of #14979, I realized that only qt 5.5+ was being tested under CI, while compatibility lists 5.2+.

Ubuntu 14.04 carries QT 5.2.1, so running it on CI is a good candidate for maintaining compatibility in an ongoing basis. Adding the change revealed some incompatibilities that have crept in.

Alternatively, we can increase the dependencies to match those carried by Ubuntu 16.04. I don't think we should claim support for dependencies we aren't testing on an ongoing basis.

Putting this forward for comment. Currently the CI run times out after build and some tests running.

Fixes #14983

Empact committed:
Do not apply attributes to lambda expressions

This was added to clang after clang 3.4. Current minimum supported version
of clang is 3.3.
https://github.com/Microsoft/clang/commit/4bfb063cae5d2dbe59778bbcd7fea4ca143c7743

Example failure:
    validation.cpp:3820:85: error: expected body of lambda expression
        if (!blocktree.LoadBlockIndexGuts(consensus_params, [this](const uint256& hash) __attribute__((exclusive_locks_required(cs_main))) { return this->InsertBlockIndex(hash); }))
                                                                                        ^
    1 error generated.
https://travis-ci.org/Empact/bitcoin/jobs/468855360

Empact committed:
bench: Implement IS_TRIVIALLY_CONSTRUCTIBLE via __has_trivial_constructor on clang

libcxx, which carries the implementation is apparently unversioned, so
this is an alternative to tracking down the specific version of clang
which switched from has_trivial_default_constructor to
is_trivially_default_constructible.
https://github.com/llvm-mirror/libcxx/commit/2fd6d25bf1758218aa71938ab343dcaefff4ffeb
https://libcxx.llvm.org/

    bench/prevector.cpp:27:21: error: no template named 'is_trivially_default_constructible' in namespace 'std'; did you mean 'has_trivial_default_constructor'?
    static_assert(!std::is_trivially_default_constructible<nontrivial_t>::value,
                   ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                        has_trivial_default_constructor
    /usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/type_traits:1211:12: note: 'has_trivial_default_constructor' declared here
        struct has_trivial_default_constructor
               ^
    bench/prevector.cpp:31:20: error: no template named 'is_trivially_default_constructible' in namespace 'std'; did you mean 'has_trivial_default_constructor'?
    static_assert(std::is_trivially_default_constructible<trivial_t>::value,
                  ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                       has_trivial_default_constructor
    /usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/type_traits:1211:12: note: 'has_trivial_default_constructor' declared here
        struct has_trivial_default_constructor
               ^
    2 errors generated.
https://travis-ci.org/Empact/bitcoin/jobs/468866070

Empact committed:
qt: Use the SLOT calling convention for QTimer::singleShot

Until Qt 5.4, singleShot did not support the new funtion-pointer-based call signature.
http://doc.qt.io/archives/qt-5.5/qtimer.html#singleShot-2

    qt/bitcoin.cpp:508:9: error: no matching function for call to 'singleShot'
            QTimer::singleShot(100, paymentServer, &PaymentServer::uiReady);
            ^~~~~~~~~~~~~~~~~~
    /usr/include/qt5/QtCore/qtimer.h:81:17: note: candidate function not viable: no known conversion from 'void (PaymentServer::*)()' to 'const char *' for 3rd argument
        static void singleShot(int msec, const QObject *receiver, const char *member);
                    ^
    /usr/include/qt5/QtCore/qtimer.h:82:17: note: candidate function not viable: requires 4 arguments, but 3 were provided
        static void singleShot(int msec, Qt::TimerType timerType, const QObject *receiver, const char *member);
                    ^
    1 error generated.
https://travis-ci.org/Empact/bitcoin/jobs/469269171

promag commented:
I was under the impression that the minimum would bump?

MarcoFalke commented on .travis.yml:
```diff
@@ -101,6 +101,15 @@ jobs:
+        HOST=x86_64-unknown-linux-gnu
+        DOCKER_NAME_TAG=ubuntu:14.04
+        PACKAGES="libglib2.0-dev libicu-dev libpng-dev python3-zmq qtbase5-dev qttools5-dev-tools protobuf-compiler libdbus-1-dev libharfbuzz-dev libprotobuf-dev"
+        DEP_OPTS="NO_QT=1 NO_UPNP=1 DEBUG=1 ALLOW_HOST_PACKAGES=1"
```
This should say `NO_DEPENDS=1` and we should use the boost libs from ubuntu (E.g. https://packages.ubuntu.com/trusty/libboost-thread-dev)?

MarcoFalke replied:
See  #15056

MarcoFalke commented on .travis.yml:
```diff
@@ -101,6 +101,15 @@ jobs:
+        PACKAGES="libglib2.0-dev libicu-dev libpng-dev python3-zmq qtbase5-dev qttools5-dev-tools protobuf-compiler libdbus-1-dev libharfbuzz-dev libprotobuf-dev"
+        DEP_OPTS="NO_QT=1 NO_UPNP=1 DEBUG=1 ALLOW_HOST_PACKAGES=1"
+        GOAL="install"
+        BITCOIN_CONFIG="--enable-zmq --with-gui=qt5 --enable-glibc-back-compat --enable-reduce-exports --enable-debug CXXFLAGS=\"-g0 -O2\""
```
Could use `--with-gui=no` for now and only fix the code (in a separate pull) when we decide to support 5.2 (as opposed to 5.4/5.5)?

Empact replied:
5.2 is listed as the minimum required version here: https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md

Empact committed:
Piecewise construct to avoid invalid construction

In CMainSignals::RegisterWithMempoolSignals running under Ubuntu 14.04
(QT 5.2), absent piecewise construction this fails to create the pair
because the argument is a connection, which is converted into a
non-copyable scoped_connection.

    validationinterface.cpp:80:186:   required from here
    /usr/include/boost/signals2/connection.hpp:234:7: error: ‘boost::signals2::scoped_connection::scoped_connection(const boost::signals2::scoped_connection&)’ is private
           scoped_connection(const scoped_connection &other);
           ^
    In file included from /usr/include/c++/4.8/utility:70:0,
                     from /usr/include/c++/4.8/algorithm:60,
                     from ./prevector.h:13,
                     from ./script/script.h:10,
                     from ./primitives/transaction.h:11,
                     from ./validationinterface.h:9,
                     from validationinterface.cpp:6:
    /usr/include/c++/4.8/bits/stl_pair.h:134:45: error: within this context
      : first(std::forward<_U1>(__x)), second(__y) { }
https://travis-ci.org/bitcoin/bitcoin/jobs/473689141#L2172

Empact commented:
Not clear why the build is failing - the reported test seems to pass:
```
2/113 - feature_maxuploadtarget.py failed, Duration: 59 s
stdout:
2018-12-31T10:43:50.480000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20181231_104336/feature_maxuploadtarget_108
2018-12-31T10:44:06.536000Z TestFramework (INFO): Peer 0 disconnected after downloading old block too many times
2018-12-31T10:44:47.537000Z TestFramework (INFO): Peer 1 able to repeatedly download new block
2018-12-31T10:44:47.640000Z TestFramework (INFO): Peer 1 disconnected after trying to download old block
2018-12-31T10:44:47.640000Z TestFramework (INFO): Advancing system time on node to clear counters...
2018-12-31T10:44:47.742000Z TestFramework (INFO): Peer 2 able to download old block
2018-12-31T10:44:47.743000Z TestFramework (INFO): Restarting nodes with -whitelist=127.0.0.1
2018-12-31T10:44:49.319000Z TestFramework (INFO): Peer still connected after trying to download old block (whitelisted)
2018-12-31T10:44:49.373000Z TestFramework (INFO): Stopping nodes
2018-12-31T10:44:49.525000Z TestFramework (INFO): Cleaning up /tmp/test_runner_₿_🏃_20181231_104336/feature_maxuploadtarget_108 on exit
2018-12-31T10:44:49.525000Z TestFramework (INFO): Tests successful
stderr:
Exception ignored in: 
TEST                                  | STATUS    | DURATION
wallet_hd.py                          | ✓ Passed  | 7 s
feature_maxuploadtarget.py            | ✖ Failed  | 59 s
```
https://travis-ci.org/bitcoin/bitcoin/jobs/473717379#L2731

Empact committed:
test: Provide test_runner TestResult functionality when status is None

Avoid e.g. sort failures by reliably returning a value for every result:
https://travis-ci.org/bitcoin/bitcoin/jobs/473943177#L2728

MarcoFalke commented:
It fails because the stderr is not empty

Empact commented:
> It fails because the stderr is not empty

Thanks, that prompted this: #15076

Empact committed:
test: Don't consider 'Exception ignored in: ' a failure

https://travis-ci.org/bitcoin/bitcoin/jobs/474288596#L2733

Empact commented:
Travis has not been running against this PR as of today. Pinged Github to raise the issue.

fanquake commented on .travis.yml:
```diff
@@ -109,6 +109,16 @@ jobs:
         BITCOIN_CONFIG="--enable-zmq --disable-wallet --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER --with-sanitizers=thread --disable-hardening --disable-asm CC=clang CXX=clang++"
 
+    - stage: test
+      name: ''x86_64 Linux  [GOAL: install]  [trusty]  [no depends, only system libs]''
```
Looks like extra quotes around the name.

Empact replied:
Ah yeah, thanks.

maflcko reviewed (commented):
You can get the travis status on https://travis-ci.org/bitcoin/bitcoin/requests. It says

` Could not parse .travis.yml `

MarcoFalke commented on test/functional/test_runner.py:
```diff
@@ -499,7 +499,10 @@ def get_next(self):
                     log_out.close(), log_err.close()
-                    if proc.returncode == TEST_EXIT_PASSED and stderr == "":
+                    # Ignore "Exception ignored in: " stderr messages
+                    # produce by python <3.6 cleanup (ubuntu < 16.04)
```
What I meant to say is to mention this only exists in python 3.4, so that we can remove it when python 3.4 support is dropped.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -3834,7 +3834,7 @@ CBlockIndex * CChainState::InsertBlockIndex(const uint256& hash)
 bool CChainState::LoadBlockIndex(const Consensus::Params& consensus_params, CBlockTreeDB& blocktree)
 {
-    if (!blocktree.LoadBlockIndexGuts(consensus_params, [this](const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return this->InsertBlockIndex(hash); }))
+    if (!blocktree.LoadBlockIndexGuts(consensus_params, [this](const uint256& hash) { AssertLockHeld(cs_main); return this->InsertBlockIndex(hash); }))
```
I think even if it compiles, the compiler couldn't enforce the lock annotations for lambdas anyway, since they could be used across compilation units?

Empact replied:
I'm not sure the implication, but I think this is good - `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` is also enforced via the `CChainState::LoadBlockIndex` declaration, and `AssertLockHeld` looks into `g_lockstack`, which is built up in the usual way. We could perhaps do without `AssertLockHeld` given the `LoadBlockIndex` declaration. Thoughts?

Empact replied:
Removing as it's also asserted in `InsertBlockIndex`

Empact replied:
Scratch that, a thread-safety-warning is issued absent this assert, thanks to the `EXCLUSIVE_LOCKS_REQUIRED` on `InsertBlockIndex`.

MarcoFalke commented:
Needs rebase

Empact committed:
Revert "travis: Compile trusty with depends for now"

Thus building against system rather than depends libs

This reverts commit fa5ce3f10ea98e5cd6bdd829b5b7b8fd2736d286.

Empact committed:
Prefer boost::optional#get_value_or over #value_or

The latter is not defined in the earliest supported version of boost,
1.47.
https://www.boost.org/doc/libs/1_47_0/libs/optional/doc/html/boost_optional/detailed_semantics.html
https://travis-ci.org/bitcoin/bitcoin/jobs/486674823

Empact commented:
Closing because we're finally green and a clean PR would be nice.
